instrumentation/cassandra/cassandra-4.4/library/src/main/java/io/opentelemetry/instrumentation/cassandra/v4_4/TracingCqlSession.java

```diff
@@ -7,199 +7,117 @@
-  }
-
-  @Override
-  public CompletionStage<PreparedStatement> prepareAsync(SimpleStatement statement) {
-    return session.prepareAsync(statement);
-  }
-
-  @Override
-  public CompletionStage<PreparedStatement> prepareAsync(String query) {
-    return session.prepareAsync(query);
-  }
-
-  @Override
-  public CompletionStage<PreparedStatement> prepareAsync(PrepareRequest request) {
-    return session.prepareAsync(request);
-  }
-
-  @Override
-  public String getName() {
-    return session.getName();
-  }
-
-  @Override
-  public Metadata getMetadata() {
-    return session.getMetadata();
-  }
-
-  @Override
-  public boolean isSchemaMetadataEnabled() {
-    return session.isSchemaMetadataEnabled();
-  }
-
-  @Override
-  public CompletionStage<Metadata> setSchemaMetadataEnabled(@Nullable Boolean newValue) {
-    return session.setSchemaMetadataEnabled(newValue);
-  }
-
-  @Override
-  public CompletionStage<Metadata> refreshSchemaAsync() {
-    return session.refreshSchemaAsync();
-  }
-
-  @Override
-  public Metadata refreshSchema() {
-    return session.refreshSchema();
-  }
-
-  @Override
-  public CompletionStage<Boolean> checkSchemaAgreementAsync() {
-    return session.checkSchemaAgreementAsync();
-  }
-
-  @Override
-  public boolean checkSchemaAgreement() {
-    return session.checkSchemaAgreement();
-  }
-
-  @Override
-  public DriverContext getContext() {
-    return session.getContext();
-  }
-
-  @Override
-  public Optional<CqlIdentifier> getKeyspace() {
-    return session.getKeyspace();
-  }
-
-  @Override
-  public Optional<Metrics> getMetrics() {
-    return session.getMetrics();
-  }
-
-  @Override
-  public CompletionStage<Void> closeFuture() {
-    return session.closeFuture();
-  }
-
-  @Override
-  public boolean isClosed() {
-    return session.isClosed();
-  }
-
-  @Override
-  public CompletionStage<Void> closeAsync() {
-    return session.closeAsync();
-  }
-
-  @Override
-  public CompletionStage<Void> forceCloseAsync() {
-    return session.forceCloseAsync();
-  }
-
-  @Override
-  public void close() {
-    session.close();
-  }
-
-  @Override
-  @Nullable
-  public <REQUEST extends Request, RESULT> RESULT execute(
-      REQUEST request, GenericType<RESULT> resultType) {
-    return session.execute(request, resultType);
-  }
+  CqlSession wrapSession(CqlSession session) {
+    if (session == null) {
+      return null;
+    }
 
-  @Override
-  public ResultSet execute(String query) {
+    List<Class<?>> interfaces = new ArrayList<>();
+    Class<?> clazz = session.getClass();
+    while (clazz != Object.class) {
+      interfaces.addAll(Arrays.asList(clazz.getInterfaces()));
+      clazz = clazz.getSuperclass();
+    }
+    return (CqlSession)
+        Proxy.newProxyInstance(
+            session.getClass().getClassLoader(),
+            interfaces.toArray(new Class<?>[0]),
+            (proxy, method, args) -> {
+              if ("execute".equals(method.getName()) && method.getParameterCount() == 1) {
```

Comment:
The `getParameterCount() == 1` check is duplicated in each `if`. If you pulled those up and returned early it could simplify a little bit.

Then, knowing that you've guarded against multiple or zero params, you can pull the `method.getParameterTypes()[0]` into a local variable to further simplify.

Reply:
I think you should view the fact that we instrument only methods that take a single parameter as a coincidence. For example the methods that we instrument delegate to https://docs.datastax.com/en/drivers/java/4.0/com/datastax/oss/driver/api/core/session/Session.html#execute-RequestT-com.datastax.oss.driver.api.core.type.reflect.GenericType- that we don't instrument, but probably should for completeness.